compiler/rustc_ast_lowering/src/delegation.rs

```diff
@@ -0,0 +1,277 @@
+use crate::{ImplTraitPosition, ResolverAstLoweringExt};
+
+use super::{ImplTraitContext, LoweringContext, ParamMode};
+
+use ast::visit::Visitor;
+use hir::def::DefKind;
+use hir::def::PartialRes;
+use hir::{BodyId, HirId};
+use rustc_ast as ast;
+use rustc_ast::*;
+use rustc_hir as hir;
+use rustc_hir::def_id::DefId;
+use rustc_middle::span_bug;
+use rustc_middle::ty::ResolverAstLowering;
+use rustc_span::symbol::kw;
+use rustc_span::{symbol::Ident, Span};
+use rustc_target::spec::abi;
+use std::iter;
+
+pub struct DelegationResults<'hir> {
+    pub body_id: hir::BodyId,
+    pub sig: hir::FnSig<'hir>,
+    pub generics: &'hir hir::Generics<'hir>,
+}
+
+impl<'hir> LoweringContext<'_, 'hir> {
+    pub fn delegation_has_self(&self, delegation: &Delegation) -> bool {
+        let res_id = self.get_delegation_res_id(delegation.id);
+        let Some(res_id) = res_id else {
+            self.tcx.sess.span_delayed_bug(
+                delegation.span,
+                "LoweringContext: couldn't resolve delegation item",
+            );
+            return false;
+        };
+
+        if let Some(local_res_id) = res_id.as_local() {
+            self.resolver.has_self.contains(&local_res_id)
+        } else {
+            match self.tcx.def_kind(res_id) {
+                DefKind::Fn => false,
+                DefKind::AssocFn => self.tcx.associated_item(res_id).fn_has_self_parameter,
+                _ => span_bug!(delegation.span, "unexpected DefKind for delegation item"),
+            }
+        }
+    }
+
+    pub fn lower_delegation(&mut self, delegation: &Delegation) -> DelegationResults<'hir> {
+        let res_id = self.get_delegation_res_id(delegation.id);
+        let Some(res_id) = res_id else {
+            self.tcx.sess.span_delayed_bug(
+                delegation.span,
+                "LoweringContext: couldn't resolve delegation item",
+            );
+            return self.generate_delegation_error(delegation);
+        };
+
+        let decl = self.lower_delegation_decl(res_id, delegation.span);
+        let sig = self.lower_delegation_sig(delegation.span, decl);
+        let body_id = self.lower_delegation_body(sig.decl, delegation);
+
+        let generics = self.lower_delegation_generics(delegation.span);
+        DelegationResults { body_id, sig, generics }
+    }
+
+    fn get_delegation_res_id(&self, node_id: NodeId) -> Option<DefId> {
+        self.resolver.get_partial_res(node_id).map(|r| r.base_res().opt_def_id()).unwrap_or(None)
+    }
+
+    fn lower_delegation_generics(&mut self, span: Span) -> &'hir hir::Generics<'hir> {
+        self.arena.alloc(hir::Generics {
+            params: self.arena.alloc_from_iter([]),
```

Comment:
```suggestion
            params: &[],
```
It doesn't make sense to allocate empty slices on arena.